### PR TITLE
DEVELOPER-3838 - Add Filtering and Facets to search

### PR DESCRIPTION
**Pull request by KamiQuasi**
PR to view new search page in live environment.

### CODE REVIEW

**Comment by paulrobinson** [comment]
@KamiQuasi this is failing to build. Seems to be an issue in the Awestruct pipeline.

**Comment by paulrobinson** [comment]
retest this please

File: bower.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "developers.redhat.com",
+  "description": "Red Hat Developers Program",
+  "main": "index.html",
+  "dependencies": {
+    "polymer": "^2.0.0",
```
Comment (LightGuard) [inline bower.json]: Thought we took out polymer.

Comment (LightGuard) [inline bower.json]: Oh, we should probably lock down the exact versions we're using too.

File: components/rhdp-search/rhdp-search-box.html [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        transition: background .2s ease-in 0s;
+        line-height: 1.2em;
+        cursor: pointer;
+        position: relative;
+        text-align: center;
+        color: #fff;
+      }
+
+      button iron-icon { display: none; }
+      @media only screen and (max-width: 768px) {
```
Comment (LightGuard) [inline components/rhdp-search/rhdp-search-box.html]: Are all of these breakpoints the same as the rest of the site?

File: components/rhdp-search/rhdp-search-filters.html [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        <div>
+        <template is="dom-repeat" items="{{filters.facets}}">
+          <template is="dom-repeat" items="{{item.items}}">
+            <template is="dom-if" if="{{item.active}}">
+              <rhdp-search-filter-item inline value="{{item.value}}" name="[[item.name]]" active="{{item.active}}"></rhdp-search-filter-item>
+            </template>
+          </template>
+        </template>
+        </div>
+        <a href="#" on-click="clearFilters">Clear Filters</a>
```
Comment (LightGuard) [inline components/rhdp-search/rhdp-search-filters.html]: Are these events hyphenated? That seems different from regular DOM.

File: components/rhdp-search/rhdp-search-query.html [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        this.set('results', qr);
+      }
+
+      checkFilters(fields, filters) {
+        var includes = false;
+        var keys = filters ? Object.keys(filters) : {};
+        var l = keys.length;
+        if(fields) {
+          for(var i=0; i<l; i++) {
+            switch (typeof filters[keys[i]]) {
```
Comment (LightGuard) [inline components/rhdp-search/rhdp-search-query.html]: The whitespace after `if` and `switch` is inconsistent, is that what we want?

File: components/rhdp-search/rhdp-search-query.html [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+              case 'object':
+                if (fields[keys[i]] && fields[keys[i]] == filters[keys[i]]) {
```
Comment (LightGuard) [inline components/rhdp-search/rhdp-search-query.html]: Inconsistent whitespace here too.

File: package.json [hunk 6]
```diff
@@ -4,24 +4,33 @@
...
   "main": "Gulpfile.js",
   "scripts": {
-    "test": "echo \"Error: no tests specified\" && exit 1"
+    "test": "karma start",
+    "bundle": "polymer-bundler --strip-comments --inline-scripts --inline-css components/rhdp-search/rhdp-search-app.html > components/rhdp-search/rhdp-search-app-build.html"
   },
   "author": "Red Hat Inc.",
   "license": "",
   "devDependencies": {
+    "bower": "^1.8.0",
```
Comment (LightGuard) [inline package.json]: This should be locked down

File: package.json [hunk 7]
```diff
@@ -4,24 +4,33 @@
...
-    "del": "2.2.0",
+    "gulp-sass": "2.1.1",
     "gulp-sourcemaps": "1.6.0",
+    "jasmine-core": "^2.6.2",
+    "karma": "^1.7.0",
+    "karma-chrome-launcher": "^2.1.1",
+    "karma-jasmine": "^1.1.0",
+    "karma-jasmine-ajax": "^0.1.13",
+    "karma-polymer": "^0.1.5",
+    "karma-spec-reporter": "0.0.31",
```
Comment (LightGuard) [inline package.json]: A bunch of these as well.

File: typescript/rhdp-search/rhdp-search-app.ts [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        facets: [
+            { name: 'CONTENT TYPE', key: 'sys_type', items: [
+                {key: 'blogpost', name: "Blog Posts", value: ['blogpost']},
+                {key: 'book', name: "Book", value: ["jbossdeveloper_book", "book"]},
+                {key: 'code', name: "Code Artifact", value: ["demo", "jbossdeveloper_archetype", "jbossdeveloper_bom"]},
+                {key: 'quickstart', name: "Quickstart", value: ['quickstart', 'quickstart_early_access']},
+                {key: 'get-started', name: "Get Started", value: ["jbossdeveloper_example"] },
+                {key: 'article-solution', name: "Knowledgebase Article / Solution", value: ["solution", "article"]},
+                {key: 'video', name: "Video", value: ["video"] }
+                ]
```
Comment (LightGuard) [inline typescript/rhdp-search/rhdp-search-app.ts]: Whitespace is inconsistent here.

File: typescript/rhdp-search/rhdp-search-filter-item.ts [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        this.setAttribute('key', this.key);
+    }
+    get active() {
+        return this._active;
+    }
+    set active(val) {
+        if(typeof val === 'string') {
+            val = true;
+        } 
+        if (this._active === val) return;
```
Comment (LightGuard) [inline typescript/rhdp-search/rhdp-search-filter-item.ts]: Why no curly braces here?

File: typescript/rhdp-search/rhdp-search-filters.ts [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+            <div class="title">${title} <a href="#" class="cancel" id="cancel">Close</a></div>
+            <div class="groups">
+            </div>
+            <div class="footer">
+            <a href="#" class="clearFilters">Clear Filters</a> 
+            <a href="#" class="applyFilters">Apply</a>
+            </div>
+        </div>`;
+    }
+    activeTemplate = (strings, title) => {
```
Comment (LightGuard) [inline typescript/rhdp-search/rhdp-search-filters.ts]: Why the longer form in some places and the shorter form in others?

File: typescript/rhdp-search/rhdp-search-result-count.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        super();
+    }
+
+    template = (strings, count, term) => {
+        return `${count} results found for ${term}`; 
+    };
+
+    connectedCallback() {
+        
+    }
```
Comment (LightGuard) [inline typescript/rhdp-search/rhdp-search-result-count.ts]: Are these empty methods needed? If so, we should document why.

**Comment by KamiQuasi** [comment]
retest this please

**Comment by KamiQuasi** [comment]
retest this please

**Comment by KamiQuasi** [comment]
retest this please

**Comment by RubyGemTSL** [comment]
I can't reproduce the test failures on my machine. It seems as though the timeouts is a wider issue (not related to this PR). I'm going to do another retest.

**Comment by RubyGemTSL** [comment]
retest this please

**Comment by RubyGemTSL** [comment]
retest this please

**Comment by KamiQuasi** [comment]
retest this please

**Comment by KamiQuasi** [comment]
retest this please

**Comment by robpblake** [comment]
retest this please

**Comment by robpblake** [comment]
retest this please

**Comment by KamiQuasi** [comment]
retest this please

**Pull request merged**